spyder/config/user.py

```diff
@@ -5,369 +5,498 @@
+        self._raw = 1 if raw_mode else 0
+        self._remove_obsolete = remove_obsolete
+        self._module_source_path = get_module_source_path('spyder')
+
+        self._defaults_folder = 'defaults'
+        self._backup_folder = 'backups'
+        self._backup_suffix = '.bak'
+        self._defaults_name_prefix = 'defaults'
+
+        # This attribute is overriding a method from cp.ConfigParser
+        self.defaults = self._check_defaults(defaults)
+
         if load:
-            # If config file already exists, it overrides Default options:
-            self.load_from_ini()
-            old_ver = self.get_version(version)
-            _major = lambda _t: _t[:_t.find('.')]
-            _minor = lambda _t: _t[:_t.rfind('.')]
+            # If config file already exists, it overrides Default options
+            previous_fpath = self.get_previous_config_fpath()
+            self._load_from_ini(previous_fpath)
+            old_version = self.get_version(version)
+            self._old_version = old_version
+
             # Save new defaults
-            self._save_new_defaults(defaults, version, subfolder)
+            self._save_new_defaults(defaults)
+
             # Updating defaults only if major/minor version is different
-            if _minor(version) != _minor(old_ver):
+            if (self._get_minor_version(version=version)
+                    != self._get_minor_version(version=old_version)):
+
                 if backup:
-                    try:
-                        shutil.copyfile(fname, "%s-%s.bak" % (fname, old_ver))
-                    except IOError:
-                        pass
-                if check_version(old_ver, '2.4.0', '<'):
-                    self.reset_to_defaults(save=False)
-                else:
-                    self._update_defaults(defaults, old_ver)
-                if check_version(old_ver, '44.1.0', '<'):
-                    run_lines = to_text_string(self.get('ipython_console',
-                                                        'startup/run_lines'))
-                    if run_lines is not NoDefault:
-                        run_lines = run_lines.replace(',', '; ')
-                        self.set('ipython_console',
-                                 'startup/run_lines', run_lines)
+                    self._make_backup(version=old_version)
+
+                self.apply_configuration_patches(old_version=old_version)
+
                 # Remove deprecated options if major version has changed
-                if remove_obsolete or _major(version) != _major(old_ver):
-                    self._remove_deprecated_options(old_ver)
+                # TODO: This might be problematic for the multi file config
+                # This needs to be called from the multiconfig manager
+                # another option should be added on init for this
+                # if (remove_obsolete or self._get_major_version(version)
+                #         != self._get_major_version(old_version)):

```

Comment:
Why is this commented? Please fix it for this PR.

Reply:
Fixed!